packages/babel-traverse/src/path/family.ts

```diff
@@ -312,6 +312,53 @@ export function getAllPrevSiblings(this: NodePath): NodePath[] {
   return siblings;
 }
 
+type digitalWithoutZero = "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9";
+type digital = "0" | digitalWithoutZero;
+
+// only support 0 to 999 (since we won't use too large number in path.get())
+type Num =
+  | `${digital}`
+  | `${digitalWithoutZero}${digital}`
+  | `${digitalWithoutZero}${digital}${digital}`;
+
+// convert "1" to 0 (string index to number index)
+type ToNumber<T extends string> = T extends Num ? 0 : T;
+
+type Pattern<Obj extends string, Prop extends string> = `${Obj}.${Prop}`;
+
+// split "body.body.1" to ["body", "body", 1]
+type Split<P extends string> = P extends Pattern<infer O, infer U>
+  ? [ToNumber<O>, ...Split<U>]
+  : [ToNumber<P>];
+
+// get all K with Node[K] is t.Node
+type NodeKeyOf<Node extends t.Node | t.Node[]> = keyof Pick<
+  Node,
+  {
+    [Key in keyof Node]-?: Node[Key] extends t.Node | t.Node[] ? Key : never;
+  }[keyof Node]
+>;
+
+// traverse the Node with tuple path ["body", "body", 1]
+// Path should be created with Split
+type Trav<
+  Node extends t.Node | t.Node[],
+  Path extends unknown[],
+> = Path extends [infer K, ...infer R]
+  ? K extends NodeKeyOf<Node>
+    ? R extends []
+      ? Node[K]
+      : // @ts-ignore ignore since TS is not smart enough
+        Trav<Node[K], R>
```

Comment:
Q: What does TS throw here? Is it the "too much recursion" error?

Reply:
Nope. It says:

```
Type 'Node[K]' does not satisfy the constraint 'Node | Node[]'.
  Type 'Node[string] | Node[number] | Node[symbol]' is not assignable to type 'Node | Node[]'.
    Type 'Node[string]' is not assignable to type 'Node | Node[]'.
      Type 'Node[string]' is not assignable to type 'Node[]'.
        Type 'Node[K]' is not assignable to type 'Node[]'.
          Type 'Node[string] | Node[number] | Node[symbol]' is not assignable to type 'Node[]'.
            Type 'Node[string]' is not assignable to type 'Node[]'.ts(2344)
```

seems that TS did not recognize that `K` is a key of `Node`.

But I tried this, which throw the same error too.

```ts
type Trav<
  Node extends t.Node | t.Node[],
  Path extends unknown[],
> = Path extends [infer K, ...infer R]
  ? K extends NodeKeyOf<Node>
    ? R extends []
      ? Node[K]
      : K extends keyof Node
      ? Trav<Node[K], R>
      : never
    : never
  : never;
```

Reply:
```suggestion
      : // @ts-expect-error TS does not infer that Node[K]'s type is correct
        Trav<Node[K], R>
```

`@ts-expect-error` will let us know when TS starts supporting this case and we can remove the error suppression comment.